File: java-api-core/src/com/kinvey/java/query/MongoQueryFilter.java
/** 
 * Copyright (c) 2013, Kinvey, Inc. All rights reserved.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 * 
 */
package com.kinvey.java.query;

import com.google.api.client.util.Key;

import java.io.Serializable;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.AbstractMap;
import java.util.Map;
import java.util.Set;

/**
 * Implementation of QueryFilter and QueryFilter.builder for Mongo
 *
 * @author mjsalinger
 * @since 2.0
 */
public class MongoQueryFilter implements QueryFilter, Serializable {

    private static final long serialVersionUID = 5635939847038496849L;
    @Key
    private LinkedHashMap<String,Object> queryFilter;

    public MongoQueryFilter() {
        queryFilter = new LinkedHashMap();
    }

    @Override
    public void put(String key, Object value) {
        queryFilter.put(key, value);
    }

    @Override
    public void put(String operation, String key, Object value) {
        LinkedHashMap<String,Object> nestedMap = new LinkedHashMap();
        nestedMap.put(operation,value);

        if (queryFilter.containsKey(key)) {
            Object currentKey = queryFilter.get(key);
            if (currentKey.getClass().equals(String.class)) {
                queryFilter.put(key, nestedMap);
            } else {
                LinkedHashMap<String,Object> currentMap = (LinkedHashMap<String,Object>) currentKey;
                currentMap.put(operation,value);
            }
        } else {
            queryFilter.put(key,nestedMap);
        }
    }

    public void put(String operation, String key, Object[] value) {
        LinkedHashMap<String,Object> nestedMap = new LinkedHashMap();
        nestedMap.put(operation,value);

        if (queryFilter.containsKey(key)) {
            Object currentKey = queryFilter.get(key);
            if (currentKey.getClass().equals(String.class)) {
                queryFilter.put(key, nestedMap);
            } else {
                LinkedHashMap<String,Object> currentMap = (LinkedHashMap<String,Object>) currentKey;
                currentMap.put(operation,value);
            }
        } else {
            queryFilter.put(key,nestedMap);
        }
    }

    @Override
    public void joinFilters(String operation, LinkedHashMap<String,Object> filterToJoin) {
        LinkedHashMap<String,Object> newFilter = new LinkedHashMap();
        LinkedHashMap<String,Object>[] combinedQueries = new LinkedHashMap[]
                {this.queryFilter, filterToJoin};
        newFilter.put(operation, combinedQueries);
        this.queryFilter = newFilter;
    }

    @Override
    public Object get(String key) {
        return queryFilter.get(key);
    }

    @Override
    public void reset() {
        queryFilter = new LinkedHashMap<String, Object>();
    }
    @Override
    public LinkedHashMap<String, Object> getFilter() {
        return this.queryFilter;
    }

    public void setFilter(LinkedHashMap<String, Object> map) {
        this.queryFilter = map;
    }

    public static class MongoQueryFilterBuilder implements QueryFilterBuilder, Serializable {

        private static final long serialVersionUID = 5635939847038496849L;

        private MongoQueryFilter query;
        private HashMap<Operators, String> operatorMap;

        public MongoQueryFilterBuilder() {
            query = new MongoQueryFilter();
            populateOperators();
        }

        public void addFilter(String operator, String key, Object value) {
            query.put(operator,key,value);
        }

        public void addFilter(String operator, String key, Object[] value) {
            query.put(operator, key, value);
        }

        @Override
        public void equals(String key, Object value) {
            query.put(key,value);
        }

        @Override
        public void resetFilter() {

        }

        public AbstractMap<String, Object> getFilterMap() {
            return (AbstractMap<String, Object>) query.queryFilter;
        }

        public void joinFilter(String operator, AbstractQuery newQuery) {
            query.joinFilters(operator, (LinkedHashMap<String, Object>) newQuery.getQueryFilterMap());
        }

        public void addLocationFilter(String field, String operator, double[] point, double distance) {
            LinkedHashMap<String,Object> locationFilter = new LinkedHashMap<String, Object>();
            Double[] mongoPoint = new Double[] {point[1],point[0]};
            locationFilter.put(operator,mongoPoint);
            if(distance > 0) {
                locationFilter.put(getOperator(Operators.MAXDISTANCE),distance);
            }

            query.put(field,locationFilter);
        }

        public void addLocationWhereFilter(String field, String operator, double[][] points) {
            LinkedHashMap<String,Object> locationFilter = new LinkedHashMap<String, Object>();
            LinkedHashMap<String, Object> withinFilter = new LinkedHashMap<String, Object>();
            Double[][] mongoPoints = new Double[points.length][2];
            for (int i=0; i < mongoPoints.length; i++) {
                mongoPoints[i][0] = points[i][1];
                mongoPoints[i][1] = points[i][0];
            }

            locationFilter.put(operator, mongoPoints);
            withinFilter.put(getOperator(Operators.WITHIN), locationFilter);
            query.put(field, withinFilter);
        }

        public void negateQuery() {
            LinkedHashMap<String,Object> currentMap = query.getFilter();
            negate(currentMap);
            query.setFilter(currentMap);
        }

        private Boolean negate(LinkedHashMap<String, Object> map) {

            boolean processParent = false;
            LinkedHashMap<String,Object> newMap = new LinkedHashMap<String, Object>();
            LinkedHashMap<String,Object> tempMap = (LinkedHashMap<String,Object>) map.clone();

            for (Map.Entry<String, Object> entry : tempMap.entrySet()) {
                String key = entry.getKey();
                Object obj = entry.getValue();
                Operators op = getOperatorKey(key);
                if (obj.getClass().equals(LinkedHashMap.class)) {
                    processParent = negate((LinkedHashMap<String,Object>) obj);
                    if (processParent) {
                        LinkedHashMap<String,Object> subMap = (LinkedHashMap<String,Object>) obj;
                        map.put(key, subMap.get("$ne"));
                    }
                }

                if (op == null && obj.getClass().equals(String.class) && !processParent) {
                    LinkedHashMap<String,Object> newNotEqualMap = new LinkedHashMap();
                    newNotEqualMap.put(getOperator(Operators.NOTEQUAL), obj);
                    map.put(key,newNotEqualMap);
                } else if (op != null) {
                    switch(op) {
                        case GREATERTHAN:
                            newMap.put("$lt", obj);
                            break;
                        case LESSTHAN :
                            newMap.put("$gt", obj);
                            break;
                        case GREATERTHANEQUAL:
                            newMap.put("$lte", obj);
                            break;
                        case LESSTHANEQUAL:
                            newMap.put("$gte", obj);
                            break;
                        case IN :
                            newMap.put("$nin", obj);
                            break;
                        case NOTIN :
                            newMap.put("$in", obj);
                            break;
                        case NOTEQUAL :
                            processParent = true;
                            break;
                        default:
                            // If not an operation that is negatable, skip.
                            break;
                    }
                }
            }
            if (!processParent) {
                map.remove("$lt");
                map.remove("$gt");
                map.remove("$gte");
                map.remove("$lte");
                map.remove("$in");
                map.remove("$nin");
                for (Map.Entry<String, Object> entry : newMap.entrySet()) {
                    map.put(entry.getKey(),entry.getValue());
                }
            }
            return processParent;
        }

        // Helper method to populate the Operators Map with the mongo operators
        private void populateOperators() {
            operatorMap = new HashMap();
            operatorMap.put(Operators.GREATERTHAN, "$gt");
            operatorMap.put(Operators.LESSTHAN, "$lt");
            operatorMap.put(Operators.GREATERTHANEQUAL, "$gte");
            operatorMap.put(Operators.LESSTHANEQUAL, "$lte");
            operatorMap.put(Operators.NOTEQUAL, "$ne");
            operatorMap.put(Operators.IN, "$in");
            operatorMap.put(Operators.ALL, "$all");
            operatorMap.put(Operators.NOTIN, "$nin");
            operatorMap.put(Operators.SIZE, "$size");
            operatorMap.put(Operators.REGEX, "$regex");
            operatorMap.put(Operators.OPTIONS, "$options");
            operatorMap.put(Operators.AND, "$and");
            operatorMap.put(Operators.OR, "$or");
            operatorMap.put(Operators.NOT, "$not");
            operatorMap.put(Operators.NEARSPHERE, "$nearSphere");
            operatorMap.put(Operators.MAXDISTANCE, "$maxDistance");
            operatorMap.put(Operators.WITHIN, "$within");
            operatorMap.put(Operators.WITHINBOX, "$box");
            operatorMap.put(Operators.WITHINPOLYGON, "$polygon");
            operatorMap.put(Operators.LOCATION, "$loc");
        }

        public String getOperator(Operators operator) {
            return operatorMap.get(operator);
        }

        public Operators getOperatorKey(String value) {
            for (Operators key : operatorMap.keySet()) {
                if (operatorMap.get(key).equals(value)) {
                    return key;
                }
            }
            return null;
        }
    }
}

File: java-api-core/src/com/kinvey/java/query/MongoQueryFilter.kt
/*
 *  Copyright (c) 2016, Kinvey, Inc. All rights reserved.
 *
 * This software is licensed to you under the Kinvey terms of service located at
 * http://www.kinvey.com/terms-of-use. By downloading, accessing and/or using this
 * software, you hereby accept such terms of service  (and any agreement referenced
 * therein) and agree that you have read, understand and agree to be bound by such
 * terms of service and are of legal age to agree to such terms with Kinvey.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 *
 */

package com.kinvey.java.query

import com.google.api.client.util.Key
import com.kinvey.java.query.QueryFilter.QueryFilterBuilder
import com.kinvey.java.query.QueryFilter.QueryFilterBuilder.Operators
import java.io.Serializable
import java.util.*
import kotlin.collections.HashMap

/**
 * Implementation of QueryFilter and QueryFilter.builder for Mongo
 *
 * @author mjsalinger
 * @since 2.0
 */
open class MongoQueryFilter : QueryFilter, Serializable {

    @Key
    override var filter: LinkedHashMap<String, Any?> = LinkedHashMap()

    override fun put(key: String, value: Any?) {
        filter[key] = value
    }

    override fun put(operation: String, key: String, value: Any?) {
        val nestedMap: LinkedHashMap<String, Any?> = LinkedHashMap()
        nestedMap[operation] = value
        if (filter.containsKey(key)) {
            val currentKey = filter[key]
            if (currentKey?.javaClass == String::class.java) {
                filter[key] = nestedMap
            } else {
                val currentMap = currentKey as LinkedHashMap<String, Any?>
                currentMap[operation] = value
            }
        } else {
            filter[key] = nestedMap
        }
    }

    override fun put(operation: String, key: String, value: Array<Any?>) {
        val nestedMap: LinkedHashMap<String, Any?> = LinkedHashMap()
        nestedMap[operation] = value
        if (filter.containsKey(key)) {
            val currentKey = filter[key]
            if (currentKey?.javaClass == String::class.java) {
                filter[key] = nestedMap
            } else {
                val currentMap = currentKey as LinkedHashMap<String, Any?>
                currentMap[operation] = value
            }
        } else {
            filter[key] = nestedMap
        }
    }

    override fun joinFilters(operation: String, filterToJoin: LinkedHashMap<String, Any?>) {
        val newFilter: LinkedHashMap<String, Any?> = LinkedHashMap()
        val combinedQueries: Array<LinkedHashMap<String, Any?>> = arrayOf(this.filter, filterToJoin)
        newFilter[operation] = combinedQueries
        this.filter = newFilter
    }

    override fun get(key: String): Any? {
        return filter[key]
    }

    override fun reset() {
        this.filter = LinkedHashMap()
    }

    open class MongoQueryFilterBuilder : QueryFilterBuilder, Serializable {

        private val query: MongoQueryFilter
        private var operatorMap: HashMap<Operators, String> = HashMap()

        override fun addFilter(operator: String, key: String, value: Any?) {
            query.put(operator, key, value)
        }

        override fun addFilter(operator: String, key: String, value: Array<Any?>?) {
            query.put(operator, key, value)
        }

        override fun equals(key: String, value: Any?) {
            query.put(key, value)
        }

        override fun resetFilter() {}
        override val filterMap: AbstractMap<String, Any?>
            get() = query.filter

        override fun joinFilter(operator: String, newQuery: AbstractQuery?) {
            query.joinFilters(operator, newQuery?.queryFilterMap as LinkedHashMap<String, Any?>)
        }

        override fun addLocationFilter(field: String, operator: String, point: DoubleArray, distance: Double) {
            val locationFilter = LinkedHashMap<String, Any>()
            val mongoPoint = arrayOf(point[1], point[0])
            locationFilter[operator] = mongoPoint
            if (distance > 0) {
                locationFilter[getOperator(Operators.MAXDISTANCE)] = distance
            }
            query.put(field, locationFilter)
        }

        override fun addLocationWhereFilter(field: String, operator: String, points: Array<DoubleArray>) {
            val locationFilter = LinkedHashMap<String, Any>()
            val withinFilter = LinkedHashMap<String, Any>()
            val mongoPoints = Array(points.size) { arrayOfNulls<Double?>(2) }
            for (i in mongoPoints.indices) {
                mongoPoints[i][0] = points[i][1]
                mongoPoints[i][1] = points[i][0]
            }
            locationFilter[operator] = mongoPoints
            withinFilter[getOperator(Operators.WITHIN)] = locationFilter
            query.put(field, withinFilter)
        }

        override fun negateQuery() {
            val currentMap = query.filter
            negate(currentMap)
            query.filter = currentMap
        }

        private fun negate(map: LinkedHashMap<String, Any?>): Boolean {
            var processParent = false
            val newMap = LinkedHashMap<String, LinkedHashMap<String, Any>>()
            val tempMap = map.clone() as LinkedHashMap<String, Any>
            val element = LinkedHashMap<String, Any>()
            for ((key, obj) in tempMap) {
                val op = getOperatorKey(key)
                if (obj.javaClass == LinkedHashMap::class.java) {
                    processParent = negate(obj as LinkedHashMap<String, Any?>)
                    if (processParent) {
                        val subMap = obj
                        map[key] = subMap["\$ne"] as Any
                    }
                }
                if (op == null && obj !is Map<*, *> && !processParent) {
                    val newNotEqualMap: LinkedHashMap<String, Any?> = LinkedHashMap()
                    newNotEqualMap[getOperator(Operators.NOTEQUAL)] = obj
                    map[key] = newNotEqualMap
                } else if (op != null) {
                    when (op) {
                        Operators.GREATERTHAN -> {
                            element["\$gt"] = obj
                            newMap["\$not"] = element
                        }
                        Operators.LESSTHAN -> {
                            element["\$lt"] = obj
                            newMap["\$not"] = element
                        }
                        Operators.GREATERTHANEQUAL -> {
                            element["\$gte"] = obj
                            newMap["\$not"] = element
                        }
                        Operators.LESSTHANEQUAL -> {
                            element["\$lte"] = obj
                            newMap["\$not"] = element
                        }
                        Operators.IN -> {
                            element["\$in"] = obj
                            newMap["\$not"] = element
                        }
                        Operators.NOTIN -> {
                            element["\$nin"] = obj
                            newMap["\$not"] = element
                        }
                        Operators.NOTEQUAL -> processParent = true
                        else -> {
                        }
                    }
                }
            }
            if (!processParent) {
                map.remove("\$lt")
                map.remove("\$gt")
                map.remove("\$gte")
                map.remove("\$lte")
                map.remove("\$in")
                map.remove("\$nin")
                for ((key, value) in newMap) {
                    map[key] = value
                }
            }
            return processParent
        }

        // Helper method to populate the Operators Map with the mongo operators
        private fun populateOperators() {
            operatorMap = HashMap()
            operatorMap[Operators.GREATERTHAN] = "\$gt"
            operatorMap[Operators.LESSTHAN] = "\$lt"
            operatorMap[Operators.GREATERTHANEQUAL] = "\$gte"
            operatorMap[Operators.LESSTHANEQUAL] = "\$lte"
            operatorMap[Operators.NOTEQUAL] = "\$ne"
            operatorMap[Operators.IN] = "\$in"
            operatorMap[Operators.ALL] = "\$all"
            operatorMap[Operators.NOTIN] = "\$nin"
            operatorMap[Operators.SIZE] = "\$size"
            operatorMap[Operators.REGEX] = "\$regex"
            operatorMap[Operators.OPTIONS] = "\$options"
            operatorMap[Operators.AND] = "\$and"
            operatorMap[Operators.OR] = "\$or"
            operatorMap[Operators.NOT] = "\$not"
            operatorMap[Operators.NEARSPHERE] = "\$nearSphere"
            operatorMap[Operators.MAXDISTANCE] = "\$maxDistance"
            operatorMap[Operators.WITHIN] = "\$within"
            operatorMap[Operators.WITHINBOX] = "\$box"
            operatorMap[Operators.WITHINPOLYGON] = "\$polygon"
            operatorMap[Operators.LOCATION] = "\$loc"
        }

        override fun getOperator(operator: Operators): String {
            return operatorMap[operator] ?: ""
        }

        fun getOperatorKey(value: String): Operators? {
            for (key in operatorMap.keys) {
                if (operatorMap[key] == value) {
                    return key
                }
            }
            return null
        }

        companion object {
            private const val serialVersionUID = 5635939847038496849L
        }

        init {
            query = MongoQueryFilter()
            populateOperators()
        }
    }

    companion object {
        private const val serialVersionUID = 5635939847038496849L
    }

    init {
        this.filter = LinkedHashMap()
    }
}